### PR TITLE
pkg: build and make ocamlformat dev-tool available

### PR DESCRIPTION
**Pull request by moyodiallo**
To avoid dune capturing a dev-tool like `ocamlformat` when using `dune build @fmt`. This PR is about to build this tool independently from dune-project dependencies, to avoid conflict. After a successful build it will be directly available whenever `dune build @fmt` or `dune fmt` is invoked. 

- [X] ocamlformat: build and make it available
- [x] prioritize `dev tools` to `PATH` when `dune.lock` dir is present in the project.
- [ ] add a cram test

The next works that would be done:
- The ocamlformat stanza `(ocamlformat <version>)` could be added in another PR in order to make this one simple. 
- A PR for `odoc` tool which will be easier, just doing the same thing as `ocamlformat` one.

### CODE REVIEW

**Review by Leonidas-from-XIV — CHANGES_REQUESTED** [review]
This is a great start! :tada: 

I've left some comments on on the code, but I think more important than that is to make sure the behaviour is what we want.

Personally I think, `ocamlformat` should not be locked when a user calls `dune pkg lock`. The issue here is that even if they never use it, they always have to pay the price for creating a lock file for ocamlformat. And if we extend this to Merlin, odoc, LSP-server, utop, the user might easily spend more time locking the dependencies of dev-tools they might never use than the dependencies of their own project.

As such I suggest changing the code to move the locking to before ocamlformat is invoked the first time, which is most likely part of the `@fmt` rule. When `@fmt` is called it would then attempt to lock & build if ocamlformat is not installed yet.

**Comment by moyodiallo** [comment]
> Personally I think, `ocamlformat` should not be locked when a user calls `dune pkg lock`. The issue here is that even if they never use it, they always have to pay the price for creating a lock file for ocamlformat. And if we extend this to Merlin, odoc, LSP-server, utop, the user might easily spend more time locking the dependencies of dev-tools they might never use than the dependencies of their own project.
> 
> As such I suggest changing the code to move the locking to before ocamlformat is invoked the first time, which is most likely part of the `@fmt` rule. When `@fmt` is called it would then attempt to lock & build if ocamlformat is not installed yet.

This is more sensible, it is the direction that we need to follow, building a tool when needed.

File: src/dune_pkg/dev_tool.ml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  ; dependencies = t.dependencies
+  ; conflicts = []
+  ; depopts = []
+  ; pins = Package_name.Map.empty
+  ; conflict_class = []
+  ; loc = Loc.none
+  }
+;;
+
+let lock_dir = Path.Source.relative (Path.Source.of_string "dune.lock") "dev_tools.lock"
```
[the remark this answers is not in the document]
Comment (moyodiallo) [reply src/dune_pkg/dev_tool.ml:21]: I suggest to do this in another PR to keep this one short and avoid having lot of changes.

**Comment by moyodiallo** [comment]
Closed to target for new PR #10647  that target `main` branch

**Pull request closed without merging**